quantized elu: graph mode handling

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40284 graph mode: add hardswish inplace handling
* **#40111 quantized elu: graph mode handling**
* #40104 quantized elu: eager mode QAT handling
* #40103 quantized elu: eager mode static handling
* #40100 quantized elu: require observation

Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

vkuzo committed:
quantized elu: graph mode handling

Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2089,6 +2089,14 @@ def test_hardswish(self):
                        .run(m.graph)
 
+    def test_elu(self):
+        data = [(torch.rand((1, 2, 5, 5), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
```
you can use `self.img_data` I think

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -20,6 +20,7 @@ std::vector<std::string> _static_quantizable_call_funcs = {
     "linear",
     "batch_norm",
     "hardswish",
+    "elu",
```
does it work if you don't add this? I'm trying to remove this list at some point

vkuzo replied:
unfortunately tests do fail if we remove this

jerryzh168 commented on torch/csrc/jit/passes/quantization/quantization_patterns.h:
```diff
@@ -675,6 +675,19 @@ graph(%a_quant, %r_scale, %r_zero_point, %r_dtype):
          %r_quant = quantized::hardswish(%a_quant, %r_scale, %r_zero_point)
          return (%r_quant) )";
 
+  // quantized::elu
+  std::string elu = R"(
+graph(%a_quant, %alpha, %scale, %input_scale, %r_scale, %r_zero_point, %r_dtype):
+         %a_dequant = aten::dequantize(%a_quant)
+         %r = aten::elu(%a_dequant, %alpha, %scale, %input_scale)
+         %r_quant = aten::quantize_per_tensor(%r, %r_scale, %r_zero_point, %r_dtype)
+         return (%r_quant) )";
+
+  std::string quantized_elu = R"(
+graph(%a_quant, %alpha, %scale, %input_scale, %r_scale, %r_zero_point, %r_dtype):
+         %r_quant = quantized::elu(%a_quant, %r_scale, %r_zero_point, %alpha, %scale, %input_scale)
+         return (%r_quant) )";
```
wait, where is the helper function for these things?
also we can quantize `aten::elu_` as well?

vkuzo replied:
yeah, the complication here is that this function has some optional arguments (the confusingly names "scale" and "input_scale" which have default values), so the quantization parameters are not the last args to this function.

TBH I don't think it's worth it to make the util function handle this case unless we have a lot more callsites - it's easier and more readable to just inline it.

vkuzo replied:
> aten::elu_

we removed the inplace version (in the first PR in this stack)

jerryzh168 replied:
OK, I think `elu_` can be quantized as well?

jerryzh168 replied:
I think it's fine to quantize `aten::elu_` to `quantized::elu`

vkuzo replied:
ah, I see.  sure.

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2089,6 +2089,14 @@ def test_hardswish(self):
 
+    def test_elu(self):
+        data = [(torch.rand((1, 2, 5, 5), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        elu = torch.nn.ELU()
```
this is not enough I think, better to test functionals and inplace variants as well

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2089,6 +2089,25 @@ def test_hardswish(self):
+        modules = [torch.nn.ELU(), torch.nn.ELU(inplace=True),
+                   FunctionalELU(True), FunctionalELU(False)]
+
+        for tracing in [True, False]:
+            for m in modules:
```
nit: use itertools to reduce indentation

jerryzh168 reviewed (approved):
Looks good

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2089,6 +2089,25 @@ def test_hardswish(self):
+            def forward(self, input):
+                return torch.nn.functional.elu(input, inplace=self.inplace)
+
+        modules = [torch.nn.ELU(), torch.nn.ELU(inplace=True),
+                   FunctionalELU(True), FunctionalELU(False)]
```
looks like this can be decomposed as [torch.nn.ELU, FunctionalELU] and [True, False]

vkuzo replied:
makes sense.  Seems like it would not matter that much since inlining it is simple - lmk if you feel strongly

jerryzh168 replied:
the pattern will be repeated, and you have tracing = [True, False] as well. I think it's better to use itertools.

jerryzh168 replied:
we can do
```python
options = itertools.product([torch.nn.ELU, FunctionalELU], [True, False], [True, False])
for M, inplace, tracing in options:
 ...
```

vkuzo replied:
no opposition to itertools, but that's not what I was targeting in my comment, it was inlining the inplace versus automating it.  I guess I don't feel that strongly about this issue either, but I'm also not sure that the benefit here is worth the cost.  This is only 2 LOC and is really easy to read.

jerryzh168 replied:
I thought about this again, it's OK to keep it this way since it's a coincidence that both of the modules has `inplace` actually

vkuzo replied:
np, the suggestion sg

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]

vkuzo committed:
Update on "quantized elu: graph mode handling"


Summary:

Adds graph mode handling for quantized elu.

Test Plan:

```
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_elu
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22075080](https://our.internmc.facebook.com/intern/diff/D22075080)

[ghstack-poisoned]
